File: Java/Decorator/CoffeeShop/src/com/ericgtkb/CondimentDecorator.java
package com.ericgtkb;

public abstract class CondimentDecorator extends Beverage {
    public abstract String getDescription();
}


// Concrete decorator classes
class Mocha extends CondimentDecorator {
    Beverage beverage;

    public Mocha(Beverage beverage) {
        this.beverage = beverage;
    }

    @Override
    public String getDescription() {
        return "Mocha " + beverage.getDescription();
    }

    @Override
    public double cost() {
        return beverage.cost() + 0.33;
    }
}


class Soy extends CondimentDecorator {
    Beverage beverage;

    public Soy(Beverage beverage) {
        this.beverage = beverage;
    }

    @Override
    public String getDescription() {
        return "Soy " + beverage.getDescription();
    }

    @Override
    public double cost() {
        return beverage.cost() + 0.50;
    }
}

File: C++/Decorator/CoffeeShop/condiment_decorator.cc
#include "condiment_decorator.h"

Mocha::Mocha(Beverage& beverage) : beverage(beverage) {}

std::string Mocha::get_description() const {
    return "Mocha" + beverage.get_description();
}

double Mocha::cost() {
    return beverage.cost() + 0.33;
}

Soy::Soy(Beverage& beverage) : beverage(beverage) {

}


std::string Soy::get_description() const {
    return "Soy " + beverage.get_description();
}

double Soy::cost() {
    return beverage.cost() + 0.50;
}